renku/service/views/__init__.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Renku service views."""
+import os
+import re
+from functools import wraps
+
+import jwt
+from flask import request, jsonify
+from git import Repo
+
+from renku.service.config import ALLOWED_EXTENSIONS, CACHE_PROJECTS_PATH, \
+    INVALID_PARAMS_ERROR_CODE, CACHE_UPLOADS_PATH, JWT_KEY, JWT_ALGORITHM
+
+is_git_url = re.compile(
+    r'((git|ssh|http(s)?)|(git@[\w\.]+))(:(//)?)([\w\.@\:/\-~]+)(\.git)(/)?',
+    flags=re.I
+)
+
+
+def allowed_file(filename):
+    """Check if filename is allowed."""
+    has_extension = '.' in filename
+    is_allowed = filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS
+    return has_extension and is_allowed
+
+
+def find_file(user, file_id):
+    """Search for cached file."""
+    for filename in os.listdir(CACHE_UPLOADS_PATH):
+        metadata = jwt.decode(filename, JWT_KEY, algorithms=[JWT_ALGORITHM])
+        if metadata['user'] == user and file_id == filename.split('.')[1]:
+            return CACHE_UPLOADS_PATH / filename
+
+
+def validate_git_url(git_url):
+    """Parse Git URL."""
+    if not git_url or not is_git_url.match(git_url):
+        return None
+
+    name = git_url.split('/')[-1].split('.')[0]
+    return name
+
+
+def validate_project_id(user, project_id):
+    """Validate if given project id is in cache."""
+    if project_id and (CACHE_PROJECTS_PATH / user / project_id).exists():
+        return CACHE_PROJECTS_PATH / user / project_id
+
+
+def repo_sync(repo_path, remote_name):
+    """Sync the repo with the default remote."""
+    repo = Repo(repo_path)
+    is_pushed = False
+
+    for remote in repo.remotes:
+        if remote.name == remote_name:
+            remote.push()
+            is_pushed = True
+
+    return is_pushed
+
+
+def requires_identity(f):
+    """Wrapper which indicates that route requires user identification."""
+    # noqa
+    @wraps(f)
+    def decorated_function(*args, **kws):
+        """Represents decorated function."""
+        user = request.headers.get('Authorization', '').split(' ')[-1]
```

Comment:
We should require a JWT from keycloak, validate it, and use the `sub` claim as the user-id.

Reply:
This proposal is quite questionable since it would break the single responsibility principle. I don't think that every service in the system should validate a token which it receives. Infrastructure should be set up in a way where the only gateway would validate a request and ensure that incoming request into the microservice is already valid and proper. Authorization header in this service is here to ensure identity and not authentication or authorization. The goal is to identify not authorize.

Reply:
Sure that's fine. But you are now using the user's token as the user-id, and I don't understand how this is expected to work? What happens when the token changes? I agree that validating it here is perhaps not the right place - but we could still require that it is a JWT with a `sub` claim which is then used for the user id.